### PR TITLE
EXE ExFont extraction.

### PR DESCRIPTION
**Pull request by 20kdc**
While this code probably isn't perfect, it's hopefully simple and hopefully works (It seems to so far).
If no ExFont file is present as a known image, one will be written to the save directory.
NOTE: This occurs even if one is already in the save directory that is not recognized as a game directory image.
This quirk is so Emscripten-based game updates work properly.
Also, adds support for loading ExFont.bmp from the save directory.

So far, only tested with "Mother: Cognitive Dissonance", but EXFONT format does not appear to vary.

### CODE REVIEW

File: src/exe_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+bool EXEReader::ResNameCheck(uint32_t i, const char * p) {
+	if (GetU16(i) != strlen(p))
+		return false;
+	while (*p) {
+		i += 2;
+		if (GetU16(i) != *p)
+			return false;
+		p++;
+	}
+	return true;
```
Comment (Ghabry) [inline src/exe_reader.cpp]: Does the code support big endian?

Comment (20kdc) [reply src/exe_reader.cpp]: Big-endian platforms or big-endian EXEs?

Comment (Ghabry) [reply src/exe_reader.cpp]: Platforms of course, there is no RPG_RT for PowerPC (luckily)

Comment (20kdc) [reply src/exe_reader.cpp]: Yes. See GetU32/GetU16/GetU8. For simplicity's sake, GetU32 is based off of GetU16 which is based off of GetU8. Thus, no endian-specific code to worry about.

**Comment by fdelapena** [comment]
Some thoughts in favour to get this merged:

This patch is not doing reverse engineering to the executable, this means, no executable code is parsed or reversed to reach them. In fact, the PE format is a documented format which contains some blocks with static data which gets parsed even by the Windows explorer to extract the icon. There are also open source programs like 7-zip which show these resources without getting any complain.

To sum up, I think there is no license violation here, as this is done in a fair and standard file format handling way.

**Comment by 20kdc** [comment]
Should I rebase & squash, or just rebase?

File: src/player.cpp [hunk 2]
```diff
@@ -695,26 +695,29 @@ void Player::CreateGameObjects() {
...
+			EXEReader exe_reader = EXEReader(exe);
+			exe_reader.GetExfont(filename);
+			exe.close();
 #ifdef EMSCRIPTEN
-		// From scene_save : Save changed file system
-		EM_ASM(
-			FS.syncfs(function(err) {
-			});
-		);
+			// From scene_save : Save changed file system
```
Comment (Ghabry) [inline src/player.cpp]: this emscripten code block looks at the wrong position, does not belong in the exe-else block

**Comment by Ghabry** [comment]
normal rebase is enough

**Comment by 20kdc** [comment]
Rebased, but it's been a while so I should double-check stuff.

File: src/player.cpp (hunk 2 above)
Comment (20kdc) [reply src/player.cpp]: It's part of the block immediately after saving a new ExFont and before loading it.

Comment (20kdc) [reply src/player.cpp]: ...Wait, no, that's a removal...

Comment (Ghabry) [reply src/player.cpp]: oh I didn't see that you added this code, thought this was our savegame code, nvermind

**Comment by 20kdc** [comment]
Things seem fine, but test for explosions

**Comment by Ghabry** [comment]
Okay devs, please give your feedback. To sum up the controversial around this PR:
Many of us are against parsing RPG_RT.exe to configure the Player. Especially auto-detecting patches via RPG_RT-parsing is highly questionable and I'm against it.
Resource parsing is a corner-case. This feels less terrible than patch detection, therefore I'm not completely against it but it still feels wrong.
Therefore: Tell me your current opinion.


@20kdc would you allow relicensing the code to MIT license? Then we could also use it in the EasyRPG/Tools repo as a standalone resource ripper (for title and exfont).

**Comment by 20kdc** [comment]
Feel free to use my PR's modifications and additions under any license you wish.
As for my opinion, it is that resource parsing is merely an extension of the project's structure - if informally - while patch detection is closer to Wine-like activity (but in a hackier manner) and by nature is going to be an incoherent mess.

**Comment by fmatthew5876** [comment]
Our objective is to emulate the game experience as faithfully as possible. If game developers are putting game content into the exe then we have no other choice but to rip it out.

So I am not against this for game content.

**Comment by carstene1ns** [comment]
The general idea of this PR is fine. However, the emscripten stuff should be removed.
Games need preprocessing anyway, so extraction of exfont/logos should be done before.
This also allows removing `ForceLoadExfont()`, which I consider a hack anyways.

**Comment by Ghabry** [comment]
@20kdc 
To sum this up: Remove the emscripten handling, just search for "RPG_RT.exe" as usual, it will silently fail.

Which means  the workflow should be:

unnamed0

**Comment by 20kdc** [comment]
...I also rebased.

File: src/cache.cpp [hunk 3]
```diff
@@ -293,6 +293,12 @@ BitmapRef Cache::Exfont() {
 
 	if (it == cache.end() || !it->second.bitmap) {
 		std::string exfont_file = FileFinder::FindImage(".", "ExFont");
+		if (exfont_file.empty()) {
+			// Extracted EXFONTs have a specific name and are only in the save directory tree.
+            // To clarify - CreateSaveDirectoryTree *must* be called to ensure the resulting directory tree is up to date.
```
Comment (carstene1ns) [inline src/cache.cpp]: whitespaceeeeeeee.

File: src/options.h [hunk 4]
```diff
@@ -57,6 +57,21 @@
...
+ * So in case of Emscripten, instead the filename used is "exfont.dll",
+ *  which makes it clear the file is only being downloaded for the EXFONT resource.
+ * The best way to handle this is of course to just have people extracting the resource themselves,
+ *  but we can't really go around expecting people to do that.
+ * However much protection that gives... debate in PR comments.
+ */
+#ifndef EMSCRIPTEN
+#define EXE_NAME "RPG_RT.exe"
+#else
+#define EXE_NAME "exfont.dll"
```
Comment (carstene1ns) [inline src/options.h]: remove.

File: src/scene_logo.cpp [hunk 5]
```diff
@@ -120,11 +120,14 @@ void Scene_Logo::OnIndexReady(FileRequestResult*) {
 	tree->SetImportantFile(true);
 	FileRequestAsync* ini = AsyncHandler::RequestFile(INI_NAME);
 	ini->SetImportantFile(true);
+	FileRequestAsync* exe = AsyncHandler::RequestFile(EXE_NAME);
+	exe->SetImportantFile(true);
```
Comment (carstene1ns) [inline src/scene_logo.cpp]: This should be reordered I think.

Comment (Ghabry) [reply src/scene_logo.cpp]: can be removed, as emscripten won't parse RPG_RT.exe and exfont loading is already below

**Comment by 20kdc** [comment]
I should probably squash my commits at some point.

**Comment by Ghabry** [comment]
Didn't want to overwrite this branch because I made so many changes that I consider doing this bad style unless discussed.

https://github.com/Ghabry/easyrpg-player/commits/exexfont

What I did: Bit of squashing and rewording.
I changed the code to only check if:
  - ExFont in the game directory and when not found
  - reads the RPG_RT.exe ExFont
In both cases the ExFont is directly read in a memory buffer, allowed simpler logic in Cache::ExFont

Used djb2_hash to detect if the ExFont is custom. Just for informational purposes.

**Comment by carstene1ns** [comment]
Well, I am ok with your changes.
